File: backend/app/api/orchestrated.py
# Built with Spec4 AI - https://spec4.ai
"""The orchestrated-subagents app's HTTP surface: the roster, and the run stream.

`GET /roster` serves two module constants -- no model, no database, no quota.
Serving it from the API rather than duplicating it in the frontend is what keeps
one list authoritative: the coordinator's decision is validated against exactly
these ids, so a frontend copy that drifted would offer a visitor a specialist
the server would refuse.

`POST /run` starts a run and streams the coordinator's delegation decision.
`POST /dispatch` is the visitor's go-ahead and streams both specialists' columns
as they fill. **They are two requests because the human decision between them is
the pattern** -- a flag on the first would have a default, and a default would
let a delegation nobody read dispatch itself. The merge event arrives in a later
phase behind the same stream.

**The decision is streamed rather than returned**, even though it is currently
the only event. That is not ceremony: the capability requires every intermediate
step be visible rather than withheld until the run completes, and a JSON
response now would have to become a stream later -- changing the client contract
at exactly the point the run gets long enough for the difference to matter.

The response is a 200 with the refusal carried as an event, following the
convention the planning app set: a run that produced something and then stopped
must not push the client's error branch and discard it. Refusals here carry a
machine-readable `outcome`, because "your question was refused", "nothing could
check your question", "the showcase is busy" and "the coordinator broke" each
call for different copy -- reword, retry, wait, retry.
"""

from __future__ import annotations

import asyncio
import json
from collections.abc import AsyncGenerator

import structlog
from fastapi import APIRouter
from pydantic import BaseModel, Field
from sse_starlette.sse import EventSourceResponse, ServerSentEvent

from backend.app.db.session import async_session_factory
from backend.app.orchestrated.presets import public_presets
from backend.app.orchestrated.roster import public_roster
from backend.app.orchestrated.runtime import VISITOR_FACING_CALL_COUNT
from backend.app.orchestrated.schemas import DelegationDecision, RosterResponse
from backend.app.orchestrated.service import begin_run, confirm_dispatch
from backend.app.services.moderation import Moderator, moderate

logger = structlog.get_logger()

router = APIRouter()

#: Longest question accepted. Matches the moderation service's own default cap,
#: so an over-long question is refused at the schema rather than after a call.
MAX_QUESTION_CHARS = 500


class RunRequest(BaseModel):
    """A question, and the preset it claims to have come from.

    `preset_id` is a *claim*, not a credential: `service.find_preset()` verifies
    it byte-matches a curated question before letting it skip moderation.
    """

    question: str = Field(..., max_length=MAX_QUESTION_CHARS)
    preset_id: str | None = None


@router.get("/api/orchestrated/roster")
async def roster() -> RosterResponse:
    """Return the fixed specialist roster and the curated preset questions.

    Both are static module constants, so this needs no warm-up and cannot fail
    for a reason worth reporting -- which is why it is a plain typed return
    rather than the `JSONResponse` the quota-spending routes in this package
    use to carry error codes.

    Returns:
        The four specialists and the curated presets.
    """
    return RosterResponse(specialists=public_roster(), presets=public_presets())


def _event(name: str, payload: dict) -> ServerSentEvent:
    """Build one named SSE event with a JSON data payload.

    Args:
        name: The event name the client dispatches on.
        payload: The JSON-serialisable body.

    Returns:
        The event, ready to yield.
    """
    return ServerSentEvent(event=name, data=json.dumps(payload))


@router.post("/api/orchestrated/run")
async def run(payload: RunRequest) -> EventSourceResponse:
    """Start a run and stream its delegation decision.

    In this phase the stream is one `delegation` event -- or one `error` event --
    and then closes. **No specialist request is issued here**: dispatch waits on
    the visitor's explicit confirmation, which arrives as a separate request in a
    later phase.

    There is no `Depends(get_db_session)` on the signature, for the reason the
    planning app's run endpoint documents: a dependency's session is bound to
    the request scope, and this response outlives the handler.

    Args:
        payload: The question and any claimed preset id.

    Returns:
        An SSE stream carrying either a `delegation` event or an `error` event
        with a machine-readable `outcome`.
    """
    return EventSourceResponse(_stream(payload, moderate))


async def _stream(
    payload: RunRequest, moderator: Moderator
) -> AsyncGenerator[ServerSentEvent, None]:
    """Drive the delegation phase and translate its outcome onto the wire.

    Args:
        payload: The validated request body.
        moderator: The moderation gate to use.

    Yields:
        One event, then the stream closes.
    """
    completed = False
    async with async_session_factory() as session:
        try:
            outcome = await begin_run(
                session,
                question=payload.question,
                preset_id=payload.preset_id,
                moderate=moderator,
            )
        except asyncio.CancelledError:
            # The visitor navigated away before the coordinator answered.
            # Nothing further is dispatched; the hold, if one was taken, is
            # swept by the expiry sweep rather than left claiming budget.
            logger.info("orchestrated_run_abandoned", phase="delegation")
            raise

        if outcome.ready and outcome.decision is not None:
            completed = True
            yield _event(
                "delegation",
                {
                    "decision_id": outcome.decision_id,
                    "chosen_specialists": [
                        specialist.value
                        for specialist in outcome.decision.chosen_specialists
                    ],
                    "rationale": outcome.decision.rationale,
                    "briefs": [
                        {
                            "specialist_id": brief.specialist_id.value,
                            "instruction": brief.instruction,
                        }
                        for brief in outcome.decision.briefs
                    ],
                    "fit_quality": outcome.decision.fit_quality.value,
                    "model_call_count": VISITOR_FACING_CALL_COUNT,
                },
            )
        else:
            yield _event(
                "error",
                {
                    "outcome": outcome.outcome.value,
                    "message": outcome.visitor_message
                    or "The run could not be started.",
                    "decision_id": outcome.decision_id,
                },
            )

    logger.info(
        "orchestrated_run_stream_closed",
        phase="delegation",
        delivered="delegation" if completed else "error",
    )


class DispatchRequest(BaseModel):
    """The visitor's explicit go-ahead to run the two chosen specialists.

    The decision travels back rather than being looked up, because nothing
    stores it: only the *hold* is persisted, keyed by `decision_id`. That makes
    this body client-controlled input, so `service.revalidate_posted_decision`
    re-checks its structure and bounds its briefs before any of it reaches a
    model -- the same treatment the planning app's run endpoint gives a posted
    plan.
    """

    decision_id: str = Field(..., min_length=1, max_length=64)
    decision: DelegationDecision
    question: str = Field(..., max_length=MAX_QUESTION_CHARS)


@router.post("/api/orchestrated/dispatch")
async def dispatch(payload: DispatchRequest) -> EventSourceResponse:
    """Run the two chosen specialists side by side and stream both columns.

    **A separate request from `/run`, and that is the human-in-the-loop gate.**
    There is no flag on the delegation call that could default to true and no
    effect that could fire on its own: dispatching requires someone to have read
    the decision and asked for it.

    Per-specialist events are emitted independently and immediately -- a status
    as each branch starts, an answer as each settles -- so a fast specialist's
    column fills while the slow one is still working, rather than both appearing
    together at the slower one's pace.

    Args:
        payload: The decision id, the decision as shown, and the question.

    Returns:
        An SSE stream of `specialist_status`, `specialist_answer` and finally
        `fan_out_complete` or `error`.
    """
    return EventSourceResponse(_dispatch_stream(payload))


async def _dispatch_stream(
    payload: DispatchRequest,
) -> AsyncGenerator[ServerSentEvent, None]:
    """Drive the fan-out and put each event on the wire as it arrives.

    Args:
        payload: The validated request body.

    Yields:
        Each dispatch event, in the order the single writer produced them.
    """
    delivered = 0
    async with async_session_factory() as session:
        stream = confirm_dispatch(
            session,
            decision_id=payload.decision_id,
            decision=payload.decision,
            question=payload.question,
        )
        try:
            async for event in stream:
                delivered += 1
                yield _event(event.name, event.payload)
        except asyncio.CancelledError:
            # Abandoning an `async for` does not close the inner generator, so
            # the specialists would keep running against a stream nobody reads.
            # Closing it explicitly is what makes the stop immediate.
            await stream.aclose()
            logger.info("orchestrated_run_abandoned", phase="dispatch")
            raise

    logger.info("orchestrated_run_stream_closed", phase="dispatch", events=delivered)

File: frontend/src/api/orchestrated.ts
// Built with Spec4 AI - https://spec4.ai
import { fetchEventSource } from '@microsoft/fetch-event-source'


/**
 * Typed client for the orchestrated-subagents endpoints.
 *
 * One client per example app, matching `planning.ts` and `chainedCalls.ts`,
 * including their error messages: `fetch` rejects identically for "the backend
 * isn't running" and "the browser blocked the response for CORS", so the
 * unreachable message names both rather than saying something unactionable.
 *
 * The roster is fetched rather than duplicated here. It is the closed set the
 * coordinator's delegation is validated against server-side, so a frontend copy
 * that drifted would offer a visitor a specialist the server would refuse.
 */

/** One member of the fixed roster, as the server publishes it. */
export interface Specialist {
  id: string
  displayName: string
  scope: string
  /** Column accent, from the design palette. */
  color: string
}

/** One curated preset question. */
export interface PresetQuestion {
  id: string
  text: string
  /**
   * The pairing a human labelled this question with.
   *
   * A label for offline evaluation, **not** a prediction the UI should show as
   * the answer — displaying it would tell the visitor what the coordinator is
   * about to decide, which is the one thing this app exists to let them watch.
   */
  expectedPairing: string[]
}

/** Response body of GET /api/orchestrated/roster. */
export interface RosterResponse {
  specialists: Specialist[]
  presets: PresetQuestion[]
}

/** One specialist's instruction for this run. */
export interface Brief {
  specialist_id: string
  instruction: string
}

/**
 * The coordinator's decision, streamed as the run's first event.
 *
 * `fit_quality` is the coordinator's own honest read on whether the question
 * suits any two of the four. `weak` is not an error — it is the app admitting
 * the pairing is approximate, which the UI shows rather than hides.
 */
export interface DelegationDecision {
  decision_id: string
  chosen_specialists: string[]
  rationale: string
  briefs: Brief[]
  fit_quality: 'strong' | 'weak'
  /** What the run costs, as told to the visitor. */
  model_call_count: number
}

/**
 * A refusal the run reported and streamed, rather than a transport failure.
 *
 * `outcome` is what to branch on: a blocked question, an exhausted hourly
 * allowance and a coordinator failure need different copy — one is fixed by
 * rewording, one by waiting, one by retrying.
 *
 * `moderation_unavailable` is deliberately not folded into `moderation_blocked`.
 * The first means nothing examined the question; telling someone their question
 * was rejected when the checker was simply down is a claim with nothing behind
 * it. It is also the state a deployment sits in with no moderation key set, so
 * it is the common one rather than the exotic one.
 */
export interface RunError {
  outcome:
    | 'moderation_blocked'
    | 'moderation_unavailable'
    | 'usage_limit_reached'
    | 'coordinator_failed'
    | 'invalid_delegation'
  message: string
  decision_id: string
}

/** How one specialist's column is doing before it has an answer. */
export interface SpecialistStatusEvent {
  specialist_id: string
  status: 'running'
}

/**
 * One specialist's settled column.
 *
 * `status` is three-valued and the third value earns its place: `timeout` means
 * the specialist was still working when the run stopped waiting, which is a
 * different thing to show — and to offer — than `failed`.
 */
export interface SpecialistAnswerEvent {
  specialist_id: string
  status: 'ok' | 'failed' | 'timeout'
  answer: string
  key_points: string[]
  error: string | null
}

/** The fan-out finished with at least one column filled. */
export interface FanOutComplete {
  decision_id: string
  survivors: string[]
  model_call_count: number
}

/**
 * A refusal on the dispatch stream.
 *
 * `refund_run` is the server saying this attempt should not count against the
 * visitor's session runs — set when both specialists failed and there is
 * nothing to show for the run.
 */
export interface DispatchError extends RunError {
  retryable: boolean
  refund_run?: boolean
}

/** One place the two specialists cannot both be acted on. */
export interface Contradiction {
  claim_a: string
  claim_b: string
  specialist_a: string | null
  specialist_b: string | null
}

/**
 * How the two independent answers relate.
 *
 * `comparable` is false when only one specialist answered — the server forces
 * it and empties every list, so a one-sided run cannot render as a comparison.
 * Contradictions reaching the client have already survived a server-side
 * traceability check: each one's quotes were found in the answer it is
 * attributed to, so a fabricated conflict never gets this far.
 */
export interface DisagreementNote {
  summary: string
  agreements: string[]
  complements: string[]
  contradictions: Contradiction[]
  comparable: boolean
}

/** The run's final output. */
export interface MergedAnswerEvent {
  decision_id: string
  text: string
  /** The specialists that actually contributed, as the server recorded them. */
  sources_used: string[]
  disagreement_note: DisagreementNote
  model_call_count: number
}

/** One received event, discriminated by its SSE event name. */
export type RunEvent =
  | { name: 'delegation'; data: DelegationDecision }
  | { name: 'error'; data: RunError }

/** One received event on the dispatch stream. */
export type DispatchEvent =
  | { name: 'specialist_status'; data: SpecialistStatusEvent }
  | { name: 'specialist_answer'; data: SpecialistAnswerEvent }
  | { name: 'fan_out_complete'; data: FanOutComplete }
  | { name: 'merged_answer'; data: MergedAnswerEvent }
  | { name: 'error'; data: DispatchError }

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL ?? 'http://localhost:8000'

/** A failure the backend described, or an unreachable one. */
export class OrchestratedRequestError extends Error {
  readonly code: string
  readonly status: number

  constructor(message: string, code: string, status: number, cause?: unknown) {
    super(message, { cause })
    this.name = 'OrchestratedRequestError'
    this.code = code
    this.status = status
  }
}

function unreachable(cause: unknown): OrchestratedRequestError {
  return new OrchestratedRequestError(
    `Could not reach the backend at ${API_BASE_URL}. Either it isn't running, ` +
      `or this page's origin (${window.location.origin}) doesn't match the ` +
      `backend's CORS_ORIGIN setting.`,
    'unreachable',
    0,
    cause,
  )
}

/**
 * Fetch the fixed specialist roster and the curated preset questions.
 *
 * Static server-side configuration: it spends no quota and calls no model, so
 * it is safe to cache indefinitely and safe to fetch on mount.
 *
 * @returns The four specialists and the curated presets.
 * @throws OrchestratedRequestError if the request cannot be made, is blocked, or returns non-2xx.
 */
export async function fetchRoster(): Promise<RosterResponse> {
  const url = `${API_BASE_URL}/api/orchestrated/roster`

  let response: Response
  try {
    response = await fetch(url)
  } catch (cause) {
    throw unreachable(cause)
  }

  if (!response.ok) {
    throw new OrchestratedRequestError(
      `The backend returned ${response.status} for ${url}.`,
      'request_failed',
      response.status,
    )
  }

  return (await response.json()) as RosterResponse
}


/** The event names the server sends. Anything else is ignored rather than crashed on. */
const RUN_EVENT_NAMES = new Set(['delegation', 'error'])

export interface StartRunOptions {
  question: string
  /**
   * The preset this question came from, if any.
   *
   * A claim, not a credential: the server re-checks that the text byte-matches
   * the stored preset before letting it skip the moderation gate.
   */
  presetId?: string | null
  /** Called once per event, in arrival order. */
  onEvent: (event: RunEvent) => void
  /** Aborts the run. The server sees the disconnect and stops working. */
  signal?: AbortSignal
}

/**
 * Start a run and stream its events back.
 *
 * `EventSource` cannot be used: it is GET-only and a run starts from a POST
 * body. Three of `fetchEventSource`'s defaults are wrong for an endpoint that
 * spends model allowance and are overridden here, for the reasons the planning
 * client documents at length — `openWhenHidden` would re-run the whole thing
 * when a backgrounded tab returns, `onerror` retries forever unless it throws,
 * and the default `onopen` reports a rejected body as a transport fault.
 *
 * @param options - The question, the optional preset claim, the per-event callback and an abort signal.
 * @returns A promise resolving when the stream closes normally.
 * @throws OrchestratedRequestError if the backend is unreachable or the run is rejected.
 */
export async function startRun({
  question,
  presetId = null,
  onEvent,
  signal,
}: StartRunOptions): Promise<void> {
  const url = `${API_BASE_URL}/api/orchestrated/run`

  await fetchEventSource(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Accept: 'text/event-stream' },
    body: JSON.stringify({ question, preset_id: presetId }),
    signal,
    openWhenHidden: true,

    async onopen(response) {
      if (response.ok) {
        return
      }
      const body: unknown = await response.json().catch(() => null)
      const detail =
        typeof body === 'object' && body !== null && 'detail' in body
          ? String((body as { detail: unknown }).detail)
          : `The backend returned ${response.status}.`
      throw new OrchestratedRequestError(detail, 'request_failed', response.status)
    },

    onmessage(message) {
      if (!RUN_EVENT_NAMES.has(message.event)) {
        return
      }
      onEvent({ name: message.event, data: JSON.parse(message.data) } as RunEvent)
    },

    onerror(cause) {
      // Rethrowing stops the library's retry loop. Returning would retry, and
      // this endpoint spends allowance.
      if (cause instanceof OrchestratedRequestError) {
        throw cause
      }
      throw unreachable(cause)
    },
  })
}

/** The event names the dispatch stream sends. */
const DISPATCH_EVENT_NAMES = new Set([
  'specialist_status',
  'specialist_answer',
  'fan_out_complete',
  'merged_answer',
  'error',
])

export interface DispatchOptions {
  /** The delegation being confirmed. Single-use: the server redeems its hold. */
  decisionId: string
  /** The decision exactly as it was shown. Re-validated server-side. */
  decision: DelegationDecision
  /** The question the decision was made for. */
  question: string
  /** Called once per event, in arrival order. */
  onEvent: (event: DispatchEvent) => void
  /** Aborts the fan-out. The server sees the disconnect and stops the branches. */
  signal?: AbortSignal
}

/**
 * Give the go-ahead and stream both specialists' columns as they fill.
 *
 * This is the human-in-the-loop gate: a second request, made only when the
 * visitor has read the delegation and asked for it. Nothing here may ever be
 * called from an effect that fires on a decision arriving.
 *
 * The same three `fetchEventSource` defaults are overridden as on the run
 * stream, and the reason bites harder here: a reopened connection on a
 * backgrounded tab would post the same `decision_id` again, which the server
 * refuses — so the visitor would lose the run rather than pay for it twice.
 *
 * @param options - The decision being confirmed, the question, a per-event callback and an abort signal.
 * @returns A promise resolving when the stream closes normally.
 * @throws OrchestratedRequestError if the backend is unreachable or the dispatch is rejected.
 */
export async function dispatchSpecialists({
  decisionId,
  decision,
  question,
  onEvent,
  signal,
}: DispatchOptions): Promise<void> {
  const url = `${API_BASE_URL}/api/orchestrated/dispatch`

  await fetchEventSource(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Accept: 'text/event-stream' },
    body: JSON.stringify({ decision_id: decisionId, decision, question }),
    signal,
    openWhenHidden: true,

    async onopen(response) {
      if (response.ok) {
        return
      }
      const body: unknown = await response.json().catch(() => null)
      const detail =
        typeof body === 'object' && body !== null && 'detail' in body
          ? String((body as { detail: unknown }).detail)
          : `The backend returned ${response.status}.`
      throw new OrchestratedRequestError(detail, 'request_failed', response.status)
    },

    onmessage(message) {
      if (!DISPATCH_EVENT_NAMES.has(message.event)) {
        return
      }
      onEvent({ name: message.event, data: JSON.parse(message.data) } as DispatchEvent)
    },

    onerror(cause) {
      if (cause instanceof OrchestratedRequestError) {
        throw cause
      }
      throw unreachable(cause)
    },
  })
}
